cli/src/main.rs

```diff
@@ -1,19 +1,235 @@
+        #[clap(long, default_value = "3001")]
+        http_port: u16,
 
-#[derive(Parser, Debug)]
+        /// The port rpc server should bind to
+        /// Default: 50051
+        /// TODO: add support to override (both here and in the development_server)
+        #[clap(long, default_value = "50051")]
+        rpc_port: u16,
+
+        /// Open the development server in your default browser after starting.
+        /// Default: true
+        /// Possible values: true, false
+        #[clap(short, long, default_value = "true")]
+        open_browser: bool,
+        // TODO: add browser override list
+    },
+    /// Stop the Mycelia development server
+    Stop {
+        /// The domain to listen on.
+        /// Default: localhost
+        /// TODO: add support to override (both here and in the development_server)
+        #[clap(short, long, default_value = "localhost")]
+        domain: String,
+
+        /// The port rpc server should bind to
+        /// Default: 50051
+        /// TODO: add support to override (both here and in the development_server)
+        #[clap(long, default_value = "50051")]
+        rpc_port: u16,
+    },
+    /// Deploy your Mycelia project
+    Deploy,
+}
+
+#[derive(Debug, Parser)]
 #[command(author, version, about, long_about = None)]
-pub struct Args {
-    /// Wasm file to invoke
-    #[arg(short, long)]
-    path: String,
+#[command(propagate_version = true)]
+struct Cli {
+    #[command(subcommand)]
+    command: Commands,
 }
 
-fn main() -> Result<(), Error> {
-    let args = Args::parse();
+fn build() -> Result<(), DynError> {
+    let cargo = env::var("CARGO").unwrap_or_else(|_| "cargo".to_string());
+    let status = Command::new(cargo)
+        .current_dir(project_root())
+        .args(&["xtask", "build"])
+        .status()?;
+
+    if !status.success() {
+        Err(format!(
+            "`cargo xtask build` failed.
 
-    let _file = File::open(args.path)?;
+Status code: {}",
+            status.code().unwrap()
+        ))?;
+    }
 
     Ok(())
 }
+
+async fn server_listening(address: String) -> Result<(), DynError> {
+    // TODO: return https://doc.rust-lang.org/stable/std/task/enum.Poll.html# ?
+    // TODO: use rand string?
+    let echo = "poll_dev_server".to_string();
+
+    let client = DevelopmentClient::connect(address.clone());
+    if let Err(e) = client.await {
+        if e.to_string() == "transport error" {
+            println!("Server not yet started");
+            return Ok(());
+        } else {
+            return Err(e.into());
+        }
+    } else {
+        // FIXME: prevent duplicate ::connect
+        let mut client = DevelopmentClient::connect(address.clone()).await?;
+        let message = EchoRequest {
+            message: echo.clone(),
+        };
+        let request = tonic::Request::new(message);
+        let response = client.echo(request).await?;
+
+        if response.into_inner() == (EchoReply { message: echo }) {
```

Comment:
is there preference to this style over just using `match`?

Reply:
just used to using if and a matter of what you are looking for: if you are looking for one very specific state ifs do the job.

but match seems better readable indeed, but I'd still resort to `_` in stead of `else`. I'll use it more often!